File: BackEndApiPlantas/src/main/java/com/example/BackEndApiPlantas/model/Imagenes.java
package com.example.BackEndApiPlantas.model;

import java.time.LocalDateTime;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Lob;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;

@Entity
@Table(name = "imagenes")
@JsonIdentityInfo(
    generator = ObjectIdGenerators.PropertyGenerator.class,
    property = "id")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Imagenes {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Lob
    private byte[] datos;

    private String nombreArchivo;
    private String tipoMime;
    private LocalDateTime fechaSubida;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "planta_id")
    @JsonIgnoreProperties({"imagenes", "usuarios", "lotes", "hibernateLazyInitializer", "handler"})
    private Planta planta;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "usuario_id", nullable = false)
    @JsonBackReference("user-imagen") // Add this annotation to match the @JsonManagedReference in Users class
    private Users usuario;

    @Enumerated(EnumType.STRING)
    private TipoImagen tipo;

    public enum TipoImagen {
        GENERAL, PERSONAL
    }

    // Getters and Setters

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public byte[] getDatos() {
        return datos;
    }

    public void setDatos(byte[] datos) {
        this.datos = datos;
    }

    public String getNombreArchivo() {
        return nombreArchivo;
    }

    public void setNombreArchivo(String nombreArchivo) {
        this.nombreArchivo = nombreArchivo;
    }

    public String getTipoMime() {
        return tipoMime;
    }

    public void setTipoMime(String tipoMime) {
        this.tipoMime = tipoMime;
    }

    public LocalDateTime getFechaSubida() {
        return fechaSubida;
    }

    public void setFechaSubida(LocalDateTime fechaSubida) {
        this.fechaSubida = fechaSubida;
    }

    public Planta getPlanta() {
        return planta;
    }

    public void setPlanta(Planta planta) {
        this.planta = planta;
    }

    public Users getUsuario() {
        return usuario;
    }

    public void setUsuario(Users usuario) {
        this.usuario = usuario;
    }

    public TipoImagen getTipo() {
        return tipo;
    }

    public void setTipo(TipoImagen tipo) {
        this.tipo = tipo;
    }

}

File: FrontEndPlantas/app/src/main/java/com/example/frontendplantas/models/Imagenes.kt
package com.example.frontendplantas.models

import com.google.gson.annotations.SerializedName
import java.time.LocalDateTime

data class Imagenes(
    @SerializedName("id")
    val id: Long? = null,

    @SerializedName("datos")
    val datos: String?, // Los datos de la imagen en formato Base64 como String (para ser enviado como JSON)

    @SerializedName("nombreArchivo")
    val nombreArchivo: String,

    @SerializedName("tipoMime")
    val tipoMime: String,

    @SerializedName("fechaSubida")
    val fechaSubida: List<Int>,

    @SerializedName("planta")
    var planta: Planta? = null, // Ahora es un objeto de tipo Planta

    @SerializedName("usuario")
    var usuario: Users? = null, // Ahora es un objeto de tipo Users

    @SerializedName("tipo")
    var tipo: TipoImagen // Enum para el tipo de imagen
) {
    // Enum que puede ser convertido directamente con Gson
    enum class TipoImagen {
        GENERAL, PERSONAL
    }
}
